### PR TITLE
New algorithm CorrectTOFAxis

### PR DESCRIPTION
**Pull request by soininen**
This PR fixes partially the issue #18091 by introducing a new algorithm `CorrectTOFAxis` which can be used to adjust the TOF axis of a workspace so that conversion from 'TOF' to 'DeltaE' will result in the elastic peak being at the zero energy transfer. The complete fix will require the removal of similar adjustments in [`LoadILLTOF`](http://docs.mantidproject.org/nightly/algorithms/LoadILLTOF-v1.html), and the usage of this algorithm together with [`GetEiMonDet`](http://docs.mantidproject.org/nightly/algorithms/GetEiMonDet-v2.html) to get the correct TOF axis for ILL workspaces. `CorrectTOFAxis` in itself is instrument/institution independent, however.

The unit tests contain some functionality common to `GetEiMonDet2Test`. This functionality was moved to WorkspaceCreationHelpers and `GetEiMonDet2Test` was refactored accordingly.

**To test:**

The script below exemplifies the algorithm's three modes of operation. Using all the modes should give the same results: shift the TOF axis slightly and update the 'Ei' and 'wavelength' sample logs. The converted workspaces should have the elastic peaks at the zero energy transfer (although in this case the raw workspace comes pretty close as well).

```python
# Adjust the input path to the Mantid build directory
inputFile = '<mantid-build-directory>/ExternalData/Testing/Data/UnitTest/ILL/IN4/084446.nxs'
raw = Load(inputFile)
# Calibrate incident energy.
EPP = FindEPP(raw)
newEi = GetEiMonDet(raw, EPP, '1-299', 0, PulseInterval=3000)
# Correct the TOF axis directly... (updates the 'Ei' and 'wavelength' sample logs)
correctedWithEPPTable = CorrectTOFAxis(raw, EPPTable=EPP, ReferenceSpectra='1-299', EFixed=newEi)
# ...by using elastic bin index...
elasticChannel = raw.run().getProperty('monitor.elasticpeak').value
correctedWithBinIndex = CorrectTOFAxis(raw, ReferenceSpectra='1-299', ElasticBinIndex=elasticChannel, EFixed=newEi)
# ... or use a reference workspace. (also updates the sample logs)
referencedRaw = CorrectTOFAxis(raw, ReferenceWorkspace=correctedWithBinIndex)
# Convert the units
convertedCorrected = ConvertUnits(correctedWithEPPTable, Target='DeltaE', EMode='Direct')
convertedRaw= ConvertUnits(raw, Target='DeltaE', EMode='Direct')
```

*The release notes were updated accordingly.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: docs/source/algorithms/TOFAxisCorrection-v1.rst [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+If *ReferenceWorkspace* is set, this algorithm copies the X axis as well as the 'Ei' and 'wavelength' sample logs to the *OutputWorkspace*. The rest of the input properties are discarded.
+
+If no *ReferenceWorkspace* is given, the algorithm calculates the average L2 distance and the average elastic peak positions. The L2 values it gets from the *InputWorkspace* while the elastic peak positions are taken from *EPPTable*. *EPPTable* should be in the format returned by the :ref:`algm-FindEPP` algorithm. The spectra over which the averages are taken should be listed in *ReferenceSpectra*. Whether this input property refers to workspace indices, spectrum numbers or detector IDs is specified by *IndexType*. Incident energy can be either specified by *IncidentEnergy* or it is taken from the sample logs of *InputWorkspace*. In case *IncidentEnergy* is specified, the 'Ei' and 'wavelength' sample logs of *OutputWorkspace* are updated accordingly.
+
```
Comment (AndreiSavici) [inline docs/source/algorithms/TOFAxisCorrection-v1.rst]: Still not clear how you modify TOF. Put a formula!

Comment (soininen) [inline docs/source/algorithms/TOFAxisCorrection-v1.rst]: Will do. The documentation needs an update, anyway.

**Review by martyngigg — COMMENTED** [review]
Can I suggest changing the name to `CorrectTOFAxis`?

It puts the doing part first like others that start `Sum`, `Calculate` etc.

**Comment by soininen** [comment]
@martyngigg I don't have any strong arguments against `CorrectTOFAxis`. It was actually in the initial list of possible names. We ended up with the current name to get this algorithm in an easier-to-spot place in the alphabetic list of the html documentation and to prevent it from vanishing in the midsts of the other Correct* algorithms. Anyway, as you say, `CorrectTOFAxis` is more in line with what is already in Mantid so I'll rename it.

**Comment by martyngigg** [comment]
@soininen Thanks. We're definitely aware of the current issue of finding algorithms in both the MantidPlot search and the documentation and I believe there should be some resource to try and tackle both next year.

File: Framework/Algorithms/src/CorrectTOFAxis.cpp [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+/// Tag for workspace indices
+const static std::string WORKSPACE_INDEX("Workspace Index");
+} // namespace IndexTypes
+
+/** A private namespace listing the properties of CorrectTOFAxis.
+ */
+namespace PropertyNames {
+const static std::string ELASTIC_BIN_INDEX("ElasticBinIndex");
+const static std::string EPP_TABLE("EPPTable");
+const static std::string INCIDENT_ENERGY("IncidentEnergy");
```
Comment (martyngigg) [inline Framework/Algorithms/src/CorrectTOFAxis.cpp]: Would it make sense to change the name of this property to `EFixed`? We have a few algorithms that can accept an E_i/E_f and we have generally settled on the property name being `EFixed` in these cases.

Comment (soininen) [inline Framework/Algorithms/src/CorrectTOFAxis.cpp]: The algorithm was meant for direct geometry TOF spectrometers, thus the choice of name here. But if calling Ei *EFixed* is a Mantid convention, then I guess I have to comply. I only wish I had realized this earlier as I have used *IncidentEnergy* in a few algorithms before...

Comment (soininen) [inline Framework/Algorithms/src/CorrectTOFAxis.cpp]: At least the two indirect instruments I have worked with (the backscattering spectrometers IN16b at ILL and SPHERES at MLZ) don't have any notion of TOF and thus there is no need to deal with the TOF axis at all. I don't know about other indirect instruments, though. The use cases I have planned for this algorithm include only direct geometry instruments but of course I cannot predict how other people might use this.

Anyway, for me this boils down to if I would like to be consistent with some algorithms I have written earlier. In hindsight I think should have used *EFixed* in them as well. I guess this is as good point as any to start using *EFixed*, so I'll make the change.

**Comment by martyngigg** [comment]
Is it expected that the conversion via bin index and via epp table give the same result?

I first ran your script and then converted `correctedWithBinIndex` to `DeltaE`. I then ran `Rebin` with `Params=2` on `convertedRaw`, `convertedCorrected` and `convertedWithBinIndex` and then `SumSpectra` on each of those to get the picture below:

correcttofaxis

It looks as if the bin index correction is 1 bin out from the epp table. Is that expected?

**Comment by soininen** [comment]
I'm not surprised if there is a bin or two difference between the epp method and giving a fixed bin index. In the test script, the epp table is build using the [FindEPP](http://docs.mantidproject.org/nightly/algorithms/FindEPP-v1.html) algorithm, which fits a Gaussian to all histograms. The average peak centre can then differ by a few bins from the 'monitor.elasticpeak' value in the sample logs. The elastic bin index in the sample logs is just something written by the instrument control software. Where this number comes from or how accurate it is, I don't know.

**Comment by martyngigg** [comment]
@soininen Thanks for clarifying that. My main concern was that there might be an off-by-one error somewhere but I hadn't realised the index came from the control software.

Happy to :shipit:

File: Framework/Algorithms/CMakeLists.txt [hunk 3]
```diff
@@ -66,6 +66,7 @@ set ( SRC_FILES
 	src/CorelliCrossCorrelate.cpp
 	src/CorrectFlightPaths.cpp
 	src/CorrectKiKf.cpp
+        src/CorrectTOFAxis.cpp
```
Comment (ianbush) [inline Framework/Algorithms/CMakeLists.txt]: These should be tabs instead of spaces to be consistent with the rest of the file.

Comment (soininen) [inline Framework/Algorithms/CMakeLists.txt]: I should know to never to edit these files in QtCreator.

**Pull request merged**